frostsnap_core/src/device.rs

```diff
@@ -254,129 +243,181 @@ impl FrostSigner {
+                self.keys.insert(key.key_id(), key.clone());
                 Ok(vec![
                     DeviceSend::ToCoordinator(DeviceToCoordinatorMessage::KeyGenAck(session_hash)),
-                    DeviceSend::ToStorage(DeviceToStorageMessage::SaveKey),
+                    DeviceSend::ToStorage(DeviceToStorageMessage::SaveKey(key.clone())),
                 ])
             }
-            _ => Err(ActionError::WrongState {
-                in_state: self.state.name(),
-                action: "keygen_ack",
-            }),
+            action_state => {
+                self.action_state = action_state;
+                Err(ActionError::WrongState {
+                    in_state: self.action_state_name(),
+                    action: "keygen_ack",
+                })
+            }
         }
     }
 
     pub fn sign_ack(&mut self) -> Result<Vec<DeviceSend>, ActionError> {
-        match self.state.clone() {
-            SignerState::AwaitingSignAck {
-                key,
+        match &self.action_state {
+            Some(SignerState::AwaitingSignAck {
+                key_id,
                 sign_task,
-                nonces,
-            } => {
+                session_nonces,
+            }) => {
+                let key = self
+                    .keys
+                    .get(key_id)
+                    .ok_or(ActionError::StateInconsistent(format!(
+                        "key {key_id} no longer exists so can't sign"
+                    )))?;
+                let secret_share = &key.secret_share;
+                let my_session_nonces = session_nonces
+                    .get(&key.share_index)
+                    .expect("already checked");
+
                 let sign_items = sign_task.sign_items();
-                let frost = frost::new_without_nonce_generation::<Sha256>();
-                let (_, my_nonce_index, my_replenish_index) =
-                    nonces.get(&key.share_index).expect("already checked");
 
-                // ⚠ Update nonce counter. Overflow would allow nonce reuse.
-                self.nonce_counter = my_nonce_index.saturating_add(sign_items.len());
+                let new_nonces = {
+                    // ⚠ Update nonce counter. Overflow would allow nonce reuse.
+                    //
+                    // hacktuallly this doesn't prevent nonce reuse. You can still re-use the nonce at
+                    // u64::MAX.
+                    //
+                    self.nonce_counter = my_session_nonces
+                        .start
+                        .saturating_add(sign_items.len() as u64);
+
+                    // This calculates the index after the last nonce the coordinator had. This is
+                    // where we want to start providing new nonces.
+                    let replenish_start = self.nonce_counter + my_session_nonces.nonces_remaining;
+                    // How many nonces we should give them from that point
+                    let replenish_amount =
```

Comment:
So the coordinator always replenishes up to `NONCE_BATCH_SIZE`.

It feels a little odd that the coordinator is keeping track of whether it needs to request more nonces, while also the device is calculating this on their side also.

What goes wrong if we only replenish the number of nonces used in this signing session? The coordinator could begin multiple sign sessions that the device never sees and run out of nonces. But couldn't the coordinator request more nonces before starting the next session?

Reply:
> So the coordinator always replenishes up to `NONCE_BATCH_SIZE`.
> 
> It feels a little odd that the coordinator is keeping track of whether it needs to request more nonces, while also the device is calculating this on their side also.
> 
> What goes wrong if we only replenish the number of nonces used in this signing session? The coordinator could begin multiple sign sessions that the device never sees and run out of nonces. But couldn't the coordinator request more nonces before starting the next session?

Do you mean send another message sent right after signing to get more nonces?
We could do that *instead* of sending nonces here but I'm not sure what the motivation is. The motivation for slipping nonces into the signature response is to avoid spontaneously sending messages while processing could be happening (and the message gets dropped because we only have one core).

Reply:
The motivation was to avoid passing the number of remaining nonces to the device. But this seems to make a bigger and more fragile mess.

The idea was to replenish nonces alongside signing as we are currently, but only replenish `sign_items.len()`.

The coordinator would `maybe_replenish_nonces` based on the size of `NONCE_BATCH_SIZE.saturating_sub(device_nonces.nonces.len()`.

I got it almost fully working, but the last step would be for the coordinator to call this replenish nonces somewhere **prior** to signing (so it works in the event that nonces were not replenished).

It also reintroduces the `NONCE_BATCH_SIZE` onto the coordinator, which as discussed could cause issues if this value were to change.

Can take a look here if curious 5ef1984f7b3b2d6bd65fb5263d3f9f358ebaa997